source/common/event/scaled_timer.cc

```diff
@@ -0,0 +1,207 @@
+#include "envoy/event/dispatcher.h"
+#include "envoy/event/timer.h"
+#include "common/event/scaled_timer.h"
+#include <algorithm>
+#include <chrono>
+
+namespace Envoy {
+namespace Event {
+
+class ScaledTimerManager::ScaledTimerImpl final : public Timer {
+public:
+  struct ActiveTimer {
+    const MonotonicTime start;
+    const std::chrono::microseconds delay;
+    const ScopeTrackedObject* object;
+  };
+
+  ScaledTimerImpl(ScaledTimerManager& manager, TimerCb callback);
+
+  ~ScaledTimerImpl() override;
+
+  /**
+   * Run the timer callback on the given dispatcher.
+   */
+  void dispatch(Dispatcher& dispatcher);
+  const absl::optional<ActiveTimer>& active() const;
+
+  // Timer impl:
+  void enableTimer(const std::chrono::milliseconds& delay,
+                   const ScopeTrackedObject* object) override;
+
+  void enableHRTimer(const std::chrono::microseconds& delay,
+                     const ScopeTrackedObject* object) override;
+
+  bool enabled() override { return active_.has_value(); }
+
+  void disableTimer() override;
+
+private:
+  ScaledTimerManager& manager_;
+  const TimerCb callback_;
+  absl::optional<ActiveTimer> active_;
+};
+
+ScaledTimerManager::ScaledTimerImpl::ScaledTimerImpl(ScaledTimerManager& manager, TimerCb callback)
+    : manager_(manager), callback_(std::move(callback)) {}
+
+ScaledTimerManager::ScaledTimerImpl::~ScaledTimerImpl() {
+  if (active_.has_value()) {
+    manager_.descheduleTimer(this);
+  }
+}
+
+void ScaledTimerManager::ScaledTimerImpl::enableTimer(const std::chrono::milliseconds& delay,
+                                                      const ScopeTrackedObject* object) {
+  if (active_.has_value()) {
+    disableTimer();
+  }
+  active_.emplace(ActiveTimer{
+      .start = manager_.dispatcher_.timeSource().monotonicTime(),
+      .delay = delay,
+      .object = object,
+  });
+  manager_.scheduleTimer(this);
+}
+
+void ScaledTimerManager::ScaledTimerImpl::enableHRTimer(const std::chrono::microseconds& delay,
+                                                        const ScopeTrackedObject* object) {
+  if (active_.has_value()) {
+    disableTimer();
+  }
+  active_.emplace(ActiveTimer{
+      .start = manager_.dispatcher_.timeSource().monotonicTime(),
+      .delay = delay,
+      .object = object,
+  });
+  manager_.scheduleTimer(this);
+}
+
+void ScaledTimerManager::ScaledTimerImpl::disableTimer() {
+  manager_.descheduleTimer(this);
+  active_.reset();
+}
+
+void ScaledTimerManager::ScaledTimerImpl::dispatch(Dispatcher& dispatcher) {
+  ASSERT(active_.has_value());
+  if (active_->object == nullptr) {
+    callback_();
+  } else {
+    ScopeTrackerScopeState scope(active_->object, dispatcher);
+    callback_();
+  }
+  active_.reset();
```

Comment:
Ordering bug: active_ could be set to a new value inside callback_, resulting in an inconsistency in the state of the timer set kept by the timer manager.  I think you want to clear active_ before entering this method.

Please add tests that cover this case.